consensus/src/pending_order_votes.rs

```diff
@@ -0,0 +1,237 @@
+};
+use std::collections::HashMap;
+
+/// Result of the order vote processing. The failure case (Verification error) is returned
+/// as the Error part of the result.
+#[derive(Debug, PartialEq, Eq)]
+pub enum OrderVoteReceptionResult {
+    /// The vote has been added but QC has not been formed yet. Return the amount of voting power
+    /// QC currently has.
+    VoteAdded(u128),
+    /// This block has just been certified after adding the vote.
+    NewLedgerInfoWithSignatures(LedgerInfoWithSignatures),
+    /// There might be some issues adding a vote
+    ErrorAddingVote(VerifyError),
+    /// Error happens when aggregating signature
+    ErrorAggregatingSignature(VerifyError),
+    /// The author of the order vote is unknown
+    UnknownAuthor(Author),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum OrderVoteStatus {
+    EnoughVotes(LedgerInfoWithSignatures),
+    NotEnoughVotes(LedgerInfoWithPartialSignatures),
+}
+
+/// A PendingVotes structure keep track of order votes for the last few rounds
+pub struct PendingOrderVotes {
+    /// Maps LedgerInfo digest to associated signatures (contained in a partial LedgerInfoWithSignatures).
+    /// Order vote status stores caches the information on whether the votes are enough to form a QC.
+    li_digest_to_votes: HashMap<HashValue /* LedgerInfo digest */, OrderVoteStatus>,
+}
+
+impl PendingOrderVotes {
+    /// Creates an empty PendingOrderVotes structure
+    pub fn new() -> Self {
+        Self {
+            li_digest_to_votes: HashMap::new(),
+        }
+    }
+
+    /// Add a vote to the pending votes
+    // TODO: Should we add any counters here?
+    pub fn insert_order_vote(
+        &mut self,
+        order_vote: &OrderVote,
+        validator_verifier: &ValidatorVerifier,
+    ) -> OrderVoteReceptionResult {
+        // derive data from order vote
+        let li_digest = order_vote.ledger_info().hash();
+
+        // obtain the ledger info with signatures associated to the order vote's ledger info
+        let status = self.li_digest_to_votes.entry(li_digest).or_insert_with(|| {
+            // if the ledger info with signatures doesn't exist yet, create it
+            OrderVoteStatus::NotEnoughVotes(LedgerInfoWithPartialSignatures::new(
+                order_vote.ledger_info().clone(),
+                PartialSignatures::empty(),
+            ))
+        });
+
+        match status {
+            OrderVoteStatus::EnoughVotes(li_with_sig) => {
+                // we already have enough votes for this ledger info
+                OrderVoteReceptionResult::NewLedgerInfoWithSignatures(li_with_sig.clone())
+            },
+            OrderVoteStatus::NotEnoughVotes(li_with_sig) => {
+                // we don't have enough votes for this ledger info yet
+                let validator_voting_power =
+                    validator_verifier.get_voting_power(&order_vote.author());
+                if validator_voting_power.is_none() {
```

Comment:
nit: this code can be just ok_or(|_| warn!(..) UnknownAuthor(..))?

Reply:
I tried ok_or earlier. ok_or converts Option to Result. And the UnknownAuthor is returned in the Error variant. Not what I wanted.